web-search-engines/langchain4j-web-search-engine-searchapi/src/main/java/dev/langchain4j/web/search/searchapi/SearchApiWebSearchEngine.java

```diff
@@ -0,0 +1,171 @@
+    	
+        
+        // conduct the search
+        final SearchApiResponse searchapiResponse = searchapiClient.search(request);        
+        
+        final List<WebSearchOrganicResult> results = searchapiResponse.getOrganicResults().stream()
+                .map(result -> WebSearchOrganicResult.from(
+                        result.getTitle(),
+                        URI.create(result.getLink()),
+                        result.getSnippet() != null ? result.getSnippet(): result.getDisplayedLink(), // for the first few runs of the query I tried, "snippet" is somehow null
+                        null, // searchapi does not return "content"
+                        toResultMetadataMap(result))) 
+                .collect(toList());
+
+        // if the search query has a direct answer, append it to the top of the organic results
+        appendAnswerToResults(searchapiResponse, results);
+        
+        final Long totalResults = Double.valueOf(searchapiResponse.getSearchInformation().get("total_results").toString()).longValue();
+    	final WebSearchInformationResult result = WebSearchInformationResult.from(
+    			getOrDefault(totalResults, Long.valueOf(results.size())), 
+    			getOrDefault(searchapiResponse.getPagination().getCurrent(), 1), 
+    			searchapiResponse.getSearchParameters());
+    	
+    	// merge the "search_information" JSON element with the "search_metadata" JSON element
+    	searchapiResponse.getSearchMetadata().putAll(searchapiResponse.getSearchInformation()); 
+        return WebSearchResults.from(searchapiResponse.getSearchMetadata(), result, results);
+    }
+
+    public static SearchApiWebSearchEngineBuilder withApiKey(final String apiKey) {
+        return builder().apiKey(apiKey);
+    }
+    
+    private static Map<String, String> toResultMetadataMap(final OrganicResult result) {
+        final Map<String, String> metadata = new HashMap<>();
+       
+        metadata.put("position", String.valueOf(result.getPosition()));
+        metadata.put("source", result.getSource());
+        metadata.put("thumbnail", result.getThumbnail());
+        return metadata;
+    }
+    
+    @SuppressWarnings("unchecked")
+	private static void appendAnswerToResults(final SearchApiResponse response, final List<WebSearchOrganicResult> results) {
+    	// searchapi.io may include 1 or more optional JSON elements in the response that directly answers the search query in the "knowledge_graph", "answer_box", etc
+        // such optional elements counts toward the "num" parameter. For instance, if "num" is set to 5 and an "inline_images" is included, then "organic_results" will contain 4 (instead of 5) items,  
+        // which is why we have to add it back to "results" below as the first item.
+    	// See also: https://api.python.langchain.com/en/latest/_modules/langchain_community/utilities/searchapi.html#SearchApiAPIWrapper
+    	
+    	String title = null, website = "https://www.searchapi.io/", description = null;
```

Comment:
what is this line for? this is not how clean Java code should look like, why again the address your already defined above as constant?

Reply:
> what is this line for? 

I tried to explain why in the preceding comment:
> searchapi.io may include 1 or more optional JSON elements in the response
> that directly answers the search query in the "knowledge_graph",
> "answer_box", etc
> such optional elements counts toward the "num" parameter. For instance, if
> "num" is set to 5 and an "inline_images" is included, then "organic_results"
> will contain 4 (instead of 5) items,
> which is why we have to add it back to "results" below as the first item.
> See also:
> https://api.python.langchain.com/en/latest/_modules/langchain_community/utilities/searchapi.html#SearchApiAPIWrapper

Please also refer to @mark-searchapi's [comment](https://github.com/langchain4j/langchain4j/issues/1132#issuecomment-2144059116) affirming my approach.

The existing code for the Tavily search engine [does similar](https://github.com/langchain4j/langchain4j/blob/2cdfb4a5f121e4ec12eeac6ce1abd0f6a8b70e8a/web-search-engines/langchain4j-web-search-engine-tavily/src/main/java/dev/langchain4j/web/search/tavily/TavilyWebSearchEngine.java#L82-L89).

No problem at all if the maintainers weigh in and ask me to remove it.


> this is not how clean Java code should look like, why again the address you already defined above as constant?

I've fixed the nits.